processor/src/memory/mod.rs

```diff
@@ -209,17 +210,43 @@ impl Memory {
         }
     }
 
-    // TEST METHODS
-    // --------------------------------------------------------------------------------------------
-
-    /// Instantiates a new processor for testing purposes.
-    #[cfg(test)]
+    /// Reads a specific address of it's value.
+    /// Unlike read() that modifies the underlying map, get_value() only attempts to read
+    /// or return None when no value exists.
     pub fn get_value(&self, addr: u64) -> Option<Word> {
         match self.trace.get(&addr) {
             Some(addr_trace) => addr_trace.last().map(|(_, value)| *value),
             None => None,
         }
     }
+
+    /// Returns all the addresses and values stored in memory.
+    pub fn get_all_values(&self) -> Vec<(u64, Word)> {
+        let mut data: Vec<(u64, [Felt; 4])> = Vec::new();
+        for (&addr, addr_trace) in self.trace.iter() {
+            let value = match addr_trace.last() {
+                Some(t) => t.1,
+                None => INIT_MEM_VALUE,
+            };
+            data.push((addr, value));
+        }
+
+        data
+    }
+
+    /// Returns values within a range of addresses.
+    pub fn get_values(&self, range: RangeInclusive<u64>) -> Vec<(u64, Word)> {
+        let mut data: Vec<(u64, Word)> = Vec::new();
+        for (&addr, addr_trace) in self.trace.range(range) {
+            let value = match addr_trace.last() {
+                Some(t) => t.1,
+                None => INIT_MEM_VALUE,
+            };
```

Comment:
If there is an entry in the map, there should always be at least one value in the vector. So, we can probably replace the `match` with `expect`. Something like: `add_trace.last().expect("empty address trace")`